File: solutions/go/tournament/1/tournament.go
package tournament

import (
	"fmt"
	"io"
	"io/ioutil"
	"sort"
	"strings"
)

type stats struct {
	name                                 string
	matches, wins, draws, losses, points int
}

func findOrCreateTeam(name string, teams *[]*stats) *stats {
	// try to find team by name
	for _, t := range *teams {
		if t.name == name {
			return t
		}
	}
	// team not found, create it
	t := &stats{name: name}
	*teams = append(*teams, t)
	return t
}

func updateStats(result string, first, second *stats) {
	switch result {
	case "win":
		first.wins++
		second.losses++
		first.points += 3
	case "draw":
		first.draws++
		second.draws++
		first.points++
		second.points++
	case "loss":
		first.losses++
		second.wins++
		second.points += 3
	}
	first.matches++
	second.matches++
}

// Tally the results of a small football competition.
func Tally(in io.Reader, out io.Writer) error {
	teams := []*stats{}

	allb, err := ioutil.ReadAll(in)
	if err != nil {
		return err
	}

	lines := strings.Split(string(allb), "\n")
	for _, line := range lines {
		if len(line) == 0 || line[0] == '#' { // skip comments and empty lines
			continue
		}
		csv := strings.Split(line, ";")
		if len(csv) < 3 || (csv[2] != "win" && csv[2] != "draw" && csv[2] != "loss") {
			return fmt.Errorf("invalid line format %q", line)
		}
		first := findOrCreateTeam(csv[0], &teams)
		second := findOrCreateTeam(csv[1], &teams)
		updateStats(csv[2], first, second)
	}

	sort.Slice(teams, func(i, j int) bool { return teams[i].name < teams[j].name })     // sort by name
	sort.Slice(teams, func(i, j int) bool { return teams[i].points > teams[j].points }) // sort by points (desc)

	// print results
	io.WriteString(out, fmt.Sprintln("Team                           | MP |  W |  D |  L |  P"))
	for _, v := range teams {
		io.WriteString(out, fmt.Sprintf("%-31s| %2d | %2d | %2d | %2d | %2d\n", v.name, v.matches, v.wins, v.draws, v.losses, v.points))
	}
	return nil
}

File: solutions/lua/tournament/1/tournament.lua
local function sorted(t)
	local retval = {}
	for k, v in pairs(t) do
		v.name = k
		retval[#retval + 1] = v
	end
	local function sort_fn(a, b) return a.p == b.p and a.name < b.name or a.p > b.p end
	table.sort(retval, sort_fn)
	return retval
end
local function formatted(t)
	local retval = { "Team                           | MP |  W |  D |  L |  P" }
	for _, r in ipairs(t) do
		local line = string.format("%-30s | %2d | %2d | %2d | %2d | %2d", r.name, r.mp, r.w, r.d, r.l, r.p)
		table.insert(retval, line)
	end
	return retval
end
return function(results)
	local data = {}
	for _, r in ipairs(results) do
		local team1, team2, result = r:match("^([^;]+);([^;]+);([^;]+)$")
		if team1 ~= nil and team2 ~= nil and (result == "win" or result == "draw" or result == "loss") then
			data[team1] = data[team1] or { mp = 0, w = 0, d = 0, l = 0, p = 0 }
			data[team2] = data[team2] or { mp = 0, w = 0, d = 0, l = 0, p = 0 }
			data[team1].mp = data[team1].mp + 1
			data[team2].mp = data[team2].mp + 1
			if result == "draw" then
				data[team1].d = data[team1].d + 1
				data[team1].p = data[team1].p + 1
				data[team2].d = data[team2].d + 1
				data[team2].p = data[team2].p + 1
			else
				if result == "loss" then
					team1, team2 = team2, team1
				end
				data[team1].w = data[team1].w + 1
				data[team1].p = data[team1].p + 3
				data[team2].l = data[team2].l + 1
			end
		end
	end
	return formatted(sorted(data))
end
